gh-101178: refactor base64.b85encode to be memory friendly

romuald opened the pull request:
## Current description

Rewrote the `base64._85encode` method logic in C, by plugging in the `binascii` module (already taking care of the bae64 methods)

By using C and a single buffer, the memory use is reduced to a minimum, addressing the initial issue.

It also greatly improves performance as a bonus:

main
```
SMALL (11 bytes): 1575427 iterations (1.27 µs per call, 115.41 ns per byte)
MEDIUM (200 bytes): 204909 iterations (9.76 µs per call, 48.80 ns per byte)
BIG (5000 bytes): 8623 iterations (231.94 µs per call, 46.39 ns per byte)
VERYBIG (500000 bytes): 81 iterations (24.69 ms per call, 49.38 ns per byte)
```

branch
```
SMALL (11 bytes): 11230718 iterations (178.08 ns per call, 16.19 ns per byte)
MEDIUM (200 bytes): 6004721 iterations (333.07 ns per call, 1.67 ns per byte)
BIG (5000 bytes): 458005 iterations (4.37 µs per call, 873.35 ps per byte)
VERYBIG (500000 bytes): 4772 iterations (419.11 µs per call, 838.22 ps per byte)
```

Script used to test: https://gist.github.com/romuald/7aeba5f40693bb351da4abe62ad7321d


## Previous description (python refactor)

**not up to date with current PR**

Refactor code to make use of generators instead of allocating 2 potentially huge lists for large datasets

Memory gain only measured using macOS and a 5Mb input.

Using `main`:

```
Before encoding
Physical footprint:         16.3M
Physical footprint (peak):  21.3M

After encoding
Physical footprint:         45.0M
Physical footprint (peak):  244.1M
```

With refactor:
```
Before encoding
Physical footprint:         14.6M
Physical footprint (peak):  19.6M

After encoding
Physical footprint:         28.5M
Physical footprint (peak):  34.4M
```

~~The execution time is more than doubled, which may not be acceptable. However the memory used is reduced by more than 90%~~
edit: changed the algorithm to be more efficient, the performance decrease now seems to be negligible

I also have no idea how (and if) I should test this

Here is the script I've used to measure the execution time, ~~the `memdebug` can probably be adapted to read `/proc/{pid}` on Linux~~
edit: updated to work on Linux too

```python
import os
import sys
import random
import hashlib
import platform
import subprocess
from time import time

from base64 import b85encode

def memdebug():
    if platform.system == "Darwin":
        if not os.environ.get("MallocStackLogging"):
            return

        res = subprocess.check_output(["malloc_history", str(os.getpid()), "-highWaterMark", "-allBySize"])

        for line in res.splitlines():
            if line.startswith(b"Physical"):
                print(line.decode())
    elif platform.system() == "Linux":
        with open(f"/proc/{os.getpid()}/status") as reader:
            for line in reader:
                if line.startswith("VmPeak:"):
                    print(line, end="")


def main():
    # use a stable input
    rnd = random.Random()
    rnd.seed(42)
    data = rnd.randbytes(5_000_000)

    memdebug()

    start = time()
    import pdb
    try:
        res = b85encode(data)
    except Exception:
        # pdb.post_mortem()
        raise
    end = time()

    memdebug()

    print("Data length:", len(data))
    print("Output length:", len(res))
    print(f"Decode time:  {end-start:.3f}s")

    h = hashlib.md5(res).hexdigest()
    print("Hashed result", h)
    assert h == "ad97e45ba085865e70f7aa05c9a31388"

    

if __name__ == '__main__':
    main()
```

<!-- gh-issue-number: gh-101178 -->
* Issue: gh-101178
<!-- /gh-issue-number -->

pitrou commented on Lib/base64.py:
```diff
@@ -299,29 +299,61 @@ def b16decode(s, casefold=False):
+    offset = 0
+    for _ in range(n1):
+        for c in unpack512(b[offset:offset+512]):
+            yield c
```
This can be simplified to:
```suggestion
        yield from unpack512(b[offset:offset+512])
```

romuald replied:
Changed (I'm not yet used to yield from ^^)

pitrou commented on Lib/test/test_base64.py:
```diff
@@ -545,6 +547,17 @@ def test_b85encode(self):
+        # since the result is to large to fit inside a test,
+        # use a hash method to validate the test
+        self.assertEqual(len(result), 784)
+        self.assertEqual(hashlib.md5(result).hexdigest(),
```
I'm not sure md5 is always available. WDYT @gpshead ?

romuald replied:
Good catch, forgot about this
According to https://docs.python.org/3/library/hashlib.html#hashlib.algorithms_guaranteed md5 may not be present, I'll switch to a sha1

romuald committed:
gh-101178: C implementation of base64._a85encode

Initially done to reduce the huge memory consumption of the previous
implementation for large inputs, and that no memory-friendly python way was
found that did not include a performance regression

This implementation also greatly improve performance in all cases

Signed-off-by: Romuald Brunet <romuald@chivil.com>

romuald committed:
Add possible regression test in test_base64

Regression was found while testing the new C implementation, when foldspaces
was used with b85encode (since a chunk could end in z without having been
folded)

picnixz commented on Modules/binascii.c:
```diff
@@ -1239,13 +1239,101 @@ binascii_b2a_qp_impl(PyObject *module, Py_buffer *data, int quotetabs,
+                value /= 85;
+            }
+            ascii_data += 5;
+        }
```
```suggestion
        if (foldnuls && value == 0) {
            *ascii_data++ = 'z';
        }
        else if (foldspaces && value == 0x20202020) {
            *ascii_data++ = 'y';
        }
        else {
            for (j = 0; j < 5 ; j++) {
                ascii_data[4 - j] = table[value % 85];
                value /= 85;
            }
            ascii_data += 5;
        }
```

picnixz commented on Modules/binascii.c:
```diff
@@ -1239,13 +1239,101 @@ binascii_b2a_qp_impl(PyObject *module, Py_buffer *data, int quotetabs,
+    // In case `i` went over the input size, we may need to shorten the output
+    const size_t overflow = (i - bin_len);
+
+    if (overflow && !pad && foldnuls && ascii_data[-1] == 'z') {
```
~~`ascii_data[-1]` won't work in C :') I think you meant `ascii_data[strlen(ascii_data) - 1]`. Alternatively I'd suggest keeping an accumulator variable that remembers the number of characters that were written.~~

romuald replied:
I was as surprised as you, but it seems to work (since I got a regression where this condition was triggered when it should not have been)

From what I understand, `ascii_data` is a pointer that is used to "write" to the buffer allocated by the writer, and its position is the next byte that the loop will write to.

So accessing `ascii_data[0]` is the byte at the current position `ascii_data[1]` the next one, and `ascii_data[-1]` the previous one.

Unless I'm abusing some part of the C standard I do not understand

picnixz replied:
No, you're right. it's actually correct per §6.5.2.1/2:

> The deﬁnition of the subscript operator [] is that E1[E2] is identical to (*((E1)+(E2))).

You just need to guarantee that it points to valid memory index. Now the only way to trigger it is if ascii_data is empty, which is only the cased when bin_len = 0 and then `overflow` would be false so no way to trigger it.

picnixz commented on Modules/binascii.c:
```diff
@@ -1239,13 +1239,101 @@ binascii_b2a_qp_impl(PyObject *module, Py_buffer *data, int quotetabs,
 }
 
+/*[clinic input]
+binascii.b2a_base85
```
It would feel weird not to have `binascii.a2b_base85` so I would suggest keeping it private for now. Ideally, `base64` should have its own C accelerator module but maybe it's an overkill.

romuald replied:
Renamed to private

That was what delayed me initially because I had no idea how to add a module dedicated to base64, I found out that it did use the binascii one only last week

romuald committed:
Review changes Modules/binascii.c

Apply suggestions

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

romuald committed:
Review fixes: update algorithm

Inspired from git source https://github.com/git/git/blob/03944513488db4a81fdb4c21c3b515e4cb260b05/base85.c#L79

This avoid checking the chunk size on every iteration and thus improves performance

romuald committed:
Further plagiate git's implementation

Since j is not unsigned anymore we can reverse the table lookup loop

picnixz commented on Modules/binascii.c:
```diff
@@ -1239,13 +1239,104 @@ binascii_b2a_qp_impl(PyObject *module, Py_buffer *data, int quotetabs,
+    size_t i = 0 ;
+    int padding = 0;
+
+    while (i < bin_len) {
```
I would also credit the git implementation for this one as it's heavily based on it.

romuald replied:
Credit added. I don't know if I should phrase it differently?

I've also added some other comments to try to explain the logic more

romuald committed:
Rename b2a_base85 as private for now

romuald committed:
Credit to git's implementation and more comments
